aidial_adapter_openai/utils/tokenizer.py

```diff
@@ -43,31 +79,33 @@ def tokens_per_message(self) -> int:
         return 3
 
     @property
-    def tokens_per_name(self) -> int:
+    def _tokens_per_request_message_name(self) -> int:
         """
         Tokens, that are counter for "name" field in message, if it's present
         """
         if self.model == "gpt-3.5-turbo-0301":
             return -1
         return 1
 
-    def calculate_request_prompt_tokens(self, messages_tokens: int):
-        """
-        Amount of tokens, that will be counted by API
-        is greater than actual sum of tokens of all messages
-        """
-        return self.TOKENS_PER_REQUEST + messages_tokens
+    def tokenize_request(
+        self, original_request: dict, messages: List[MessageType]
+    ) -> int:
+        tokens = self.TOKENS_PER_REQUEST
 
-    def calculate_prompt_tokens(self, messages: List[MessageType]) -> int:
-        return self.calculate_request_prompt_tokens(
-            messages_tokens=sum(map(self.calculate_message_tokens, messages))
-        )
+        if original_request.get("function_call") != "none":
+            for func in original_request.get("function") or []:
+                tokens += self._tokenize_object(func)
+
+        if original_request.get("tool_choice") != "none":
+            for tool in original_request.get("tools") or []:
+                tokens += self._tokenize_object(tool.get("function"))
 
-    def available_message_tokens(self, max_prompt_tokens: int):
-        return max_prompt_tokens - self.TOKENS_PER_REQUEST
+        tokens += sum(map(self.tokenize_request_message, messages))
+
+        return tokens
 
     @abstractmethod
-    def calculate_message_tokens(self, message: MessageType) -> int:
+    def tokenize_request_message(self, message: MessageType) -> int:
```

Comment:
I think it would be easy to confuse `tokenize_request` and `tokenize_request_message` methods

Maybe let's name it just `tokenize_messsage`?

Reply:
We need to distinguish response message from a request message, that's why I introduced
`_tokenize_response_message` and `tokenize_request_message`.

`tokenize_message` doesn't help to make the distinction.